test/e2e/tests/phishing-controller/phishing-detection.spec.js

```diff
@@ -297,4 +330,151 @@ describe('Phishing Detection', function () {
+     *
+     * @param {string} name - The name of the header to set.
+     * @param {string} value - The value of the header to set.
+     * @param {typeof redirectableStatusCodes[number]} code - The status code
+     * @returns {void}
+     */
+    function handleRequests(name, value, code) {
+      server.once('request', async function (_request, response) {
+        response.setHeader(name, value).writeHead(code).end(`
+        <!doctype html>
+        <html>
+          <head>
+            <meta http-equiv="Refresh" content="0;url="${destination}"/>
+            <title>Phishing test</title>
+
+            <script>
+              // this script should not run.
+              // it is meant to test for regressions in our redirect
+              // protection due to changes in either MetaMask or browsers.
+              document.location.href = "${destination}";
+              alert("trying to prevent phishing protection");
+              while(true){}
+            </script>
+          </head>
+          <body>
+            <h1>Redirecting...</h1>
+          </body>
+        </html>
+        `);
+      });
+    }
+
+    before('Set up redirect server', async function () {
+      const { promise, resolve, reject } = createDeferredPromise();
+      server = createServer();
+      server.listen(0, blocked, resolve);
+      server.on('error', reject);
+      await promise;
+    });
+    after('Shut down redirect server', async function () {
+      if (server) {
+        const { promise, resolve } = createDeferredPromise();
+        server.close(() => resolve());
+        // We need to close all connections to stop the server quickly
+        // Otherwise it takes a few seconds for it to close
+        server.closeAllConnections();
+        await promise;
+      }
+    });
+    const deferredTestSuite = createDeferredPromise();
+    before('Set up fixtures', async function () {
+      /**
+       * @type {{ promise: Promise<Driver>, resolve: (driver: Driver) => void } | undefined
+       */
+      const { promise, resolve } = createDeferredPromise();
+      fixturePromise = withFixtures(
+        {
+          fixtures: new FixtureBuilder().build(),
+          ganacheOptions: defaultGanacheOptions,
+          title: this.test.fullTitle(),
+          testSpecificMock: async (mockServer) => {
+            await setupPhishingDetectionMocks(mockServer, {
+              blockProvider: BlockProvider.MetaMask,
+              blocklist: [blocked],
+            });
+          },
+        },
+        async (fixtures) => {
+          resolve(fixtures.driver); // resolve this `beforeEach`
+          await deferredTestSuite.promise; // now wait for all tests to complete
+        },
+      );
+      driver = await promise;
+
+      // required to ensure MetaMask is fully started before running tests
+      // if we had a way of detecting when the offscreen/background were ready
+      // we could remove this
+      await unlockWallet(driver);
```

Comment:
sad day.

I couldn't easily set up tests to wait until the background/service_worker was fully ready, so I had to fall back to the `unlockWallet` strategy.

Note: the wallet does NOT have to be unlocked for the protections to kick in.

Reply:
I've figured out a way in a separate PR meant to fix a flaky test.

@HowardBraham figured out a way to set up a socket connection from the extension to the test runner in https://github.com/MetaMask/metamask-extension/pull/25362, so we might be able to use that (which I think will be even better than my other solution)!